### ISSUE TITLE
agent installation does hang indefinitely from system class loader

### ISSUE THREAD

**Issue by Vampire** [raphw/byte-buddy#845]
I'm currently toying around with ByteBuddy and tried the following.
I have the system property `-Djava.system.class.loader=TestClassLoader` set.
`TestClassLoader` looks like
```java
import net.bytebuddy.ByteBuddy;
import net.bytebuddy.agent.ByteBuddyAgent;
import net.bytebuddy.dynamic.ClassFileLocator;
import net.bytebuddy.implementation.StubMethod;
import net.bytebuddy.pool.TypePool;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicReference;

import static net.bytebuddy.dynamic.loading.ClassLoadingStrategy.Default.INJECTION;
import static net.bytebuddy.matcher.ElementMatchers.named;

public class TestClassLoader extends ClassLoader {
  private final AtomicReference<ClassFileLocator> classFileLocator = new AtomicReference<>();

  public TestClassLoader(ClassLoader parent) {
    super(parent);
  }

  public ClassFileLocator getClassFileLocator() {
    return classFileLocator.updateAndGet(cfl -> {
      if (cfl != null) {
        return cfl;
      }
      ByteBuddyAgent.install();
      return new ClassFileLocator.Compound(
        ClassFileLocator.AgentBased.fromInstalledAgent(getParent()),
        ClassFileLocator.ForClassLoader.of(getParent()));
    });
  }

  @Override
  protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
    synchronized (getClassLoadingLock(name)) {
      Class<?> c = findLoadedClass(name);
      if (c == null) {
        if (!"Testee".equals(name)) {
          try {
            c = getParent().loadClass(name);
          } catch (ClassNotFoundException e) {
            // maybe an agent class, try with own strategy
          }
        }

        if (c == null) {
          try {
            ClassFileLocator classFileLocator = getClassFileLocator();
            TypePool.Resolution resolution = TypePool.Default.of(classFileLocator).describe(name);
            if (resolution.isResolved()) {
              c = new ByteBuddy()
                .redefine(resolution.resolve(), classFileLocator)
                .method(named("doIt"))
                .intercept(StubMethod.INSTANCE)
                .make()
                .load(this, INJECTION)
                .getLoaded();
            }
          } catch (Exception e) {
            e.printStackTrace();
          }
        }

        if (c == null) {
          c = getParent().loadClass(name);
        }
      }
      if (resolve) {
        resolveClass(c);
      }
      return c;
    }
  }

  @Override
  protected Class<?> findClass(String name) throws ClassNotFoundException {
    return loadClass(name);
  }

  /**
   * Called by the VM to support dynamic additions to the class path
   *
   * @see java.lang.instrument.Instrumentation#appendToSystemClassLoaderSearch
   */
  void appendToClassPathForInstrumentation(String path) throws IOException {
    ClassFileLocator instrumentationClassFileLocator = ClassFileLocator.ForJarFile.of(new File(path));
    classFileLocator.updateAndGet(cfl -> new ClassFileLocator.Compound(
      getClassFileLocator(),
      instrumentationClassFileLocator));
  }
}
```
`Testee` is just
```java
public class Testee {
    public void doIt() {
        System.out.println("Done");
    }
}
```
and the main class
```java
public class Test {
    public static void main(String[] args) {
        if (!"TestClassLoader".equals(System.getProperty("java.system.class.loader"))) {
            throw new RuntimeException("Please set system property 'java.system.class.loader' to 'TestClassLoader'");
        }
        new Testee().doIt();
    }
}
```

If you now run the main method, as soon as the `Testee` class is loaded, `getClassFileLocator()` is called and the lazy initialization should happen.
But the `ByteBuddyAgent.install()` line hangs indefinitely.
Is this expected or a bug?
If I use the same class loader code, but not have it as system class loader it works properly.
Meaning if I do in the main method `Class.forName("Testee", false, new TestClassLoader(getSystemClassLoader()))` it works properly, installing the agent and continuing.

**Comment by raphw** [comment]
Java agents are loaded on the system loader, I assume that there are two threads hanging up on class loading. Did you take a thread dump? I assume that Byte Buddy wants to instrument a class which requires loading its own classes where the class loader already got locked by another thread for class loading.

I assume that you can solve this problem by calling `ClassLoader.registerParallelCapable()` from a static initializer block.

**Comment by Vampire** [comment]
Indeed that did the trick which I just found out too before I have seen your answer. :-)
The agent installation was done lazily from within the class loading, so without the parallel capability, the whole class loader instance was locked.
Attaching an agent seems to try to load the agent class on the system class loader, but from a different thread and without a breakpoint at the first line of `loadClass` being hit which is probably why I initially didn't get it.
After registering the class loader as parallel capable it works as expected, thanks.


One more point though, I just had to kill 20 Java processes from the external attacher strategy as I tested with Java 8 and 13 so see if there are differences (there are not).
They all hung at `sun.tools.attach.VirtualMachineImpl#connectPipe` according to an attached IntelliJ and waited for the attaching in the target VM to conclude, even though the target VM was already killed and no longer available.

I'm not sure whether this is a JVM bug, or (or and) whether you maybe can have some timeout there.

**Comment by raphw** [comment]
This should not happen but the attach API is fairly shacky when it comes to these things. A broken pipe can always hang up a JVM but there is a timeout that should end these processes at some point.

**Comment by Vampire** [comment]
Well, they were not after hours as far as I remember.

Btw. for reference of future users, marking the class loader as parallel capable was not enough.
I also had to separate out the additional instrumentation classpath.
I tried that first as in the orignal code the class loader indeed was marked as parallel capable but it still hung.
I just stripped away the parallel capabel for a minimal reproducer, as it still hung without it.
So this version in the end worked:

```java
import net.bytebuddy.ByteBuddy;
import net.bytebuddy.agent.ByteBuddyAgent;
import net.bytebuddy.agent.builder.AgentBuilder;
import net.bytebuddy.agent.builder.AgentBuilder.RedefinitionStrategy;
import net.bytebuddy.dynamic.ClassFileLocator;
import net.bytebuddy.dynamic.ClassFileLocator.Resolution;
import net.bytebuddy.implementation.StubMethod;
import net.bytebuddy.pool.TypePool;

import java.io.File;
import java.io.IOException;
import java.lang.instrument.Instrumentation;
import java.util.concurrent.atomic.AtomicReference;

import static net.bytebuddy.agent.builder.AgentBuilder.RedefinitionStrategy.RETRANSFORMATION;
import static net.bytebuddy.dynamic.loading.ClassLoadingStrategy.Default.INJECTION;
import static net.bytebuddy.matcher.ElementMatchers.is;
import static net.bytebuddy.matcher.ElementMatchers.named;

public class TestClassLoader extends ClassLoader {
  static {
    ClassLoader.registerAsParallelCapable();
  }

  private final AtomicReference<ClassFileLocator> classFileLocator = new AtomicReference<>();
  private final AtomicReference<ClassFileLocator> instrumentationClassFileLocator = new AtomicReference<>();

  public TestClassLoader(ClassLoader parent) {
    super(parent);
    instrumentationClassFileLocator.set(ClassFileLocator.ForClassLoader.of(parent));
  }

  public ClassFileLocator getClassFileLocator() {
    return classFileLocator.updateAndGet(cfl -> {
      if (cfl != null) {
        return cfl;
      }
      ByteBuddyAgent.install();
      return new ClassFileLocator.Compound(
        ClassFileLocator.AgentBased.fromInstalledAgent(getParent()),
        ClassFileLocator.ForClassLoader.of(getParent()));
    });
  }

  @Override
  protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
    synchronized (getClassLoadingLock(name)) {
      Class<?> c = findLoadedClass(name);
      if (c == null) {
        if (!("Test".equals(name) || "Testee".equals(name))) {
          try {
            c = getParent().loadClass(name);
          } catch (ClassNotFoundException e) {
            // maybe an agent class, try with own strategy
          }
        }

        if ((c == null) && !("Test".equals(name) || "Testee".equals(name))) {
          try {
            ClassFileLocator classFileLocator = instrumentationClassFileLocator.get();
            Resolution resolution = classFileLocator.locate(name);
            if (resolution.isResolved()) {
              byte[] classBytes = resolution.resolve();
              c = defineClass(name, classBytes, 0, classBytes.length);
            }
          } catch (Exception e) {
            e.printStackTrace();
          }
        }

        if (c == null) {
          try {
            ClassFileLocator classFileLocator = getClassFileLocator();
            TypePool.Resolution resolution = TypePool.Default.of(classFileLocator).describe(name);
            if (resolution.isResolved()) {
              c = new ByteBuddy()
                .redefine(resolution.resolve(), classFileLocator)
                .method(named("doIt"))
                .intercept(StubMethod.INSTANCE)
                .make()
                .load(this, INJECTION)
                .getLoaded();
            }
          } catch (Exception e) {
            e.printStackTrace();
          }
        }

        if (c == null) {
          c = getParent().loadClass(name);
        }
      }
      if (resolve) {
        resolveClass(c);
      }
      return c;
    }
  }

  @Override
  protected Class<?> findClass(String name) throws ClassNotFoundException {
    return loadClass(name);
  }

  /**
   * Called by the VM to support dynamic additions to the class path
   *
   * @see java.lang.instrument.Instrumentation#appendToSystemClassLoaderSearch
   */
  void appendToClassPathForInstrumentation(String path) throws IOException {
    ClassFileLocator instrumentationClassFileLocator = ClassFileLocator.ForJarFile.of(new File(path));
    this.instrumentationClassFileLocator.updateAndGet(cfl ->
            new ClassFileLocator.Compound(cfl, instrumentationClassFileLocator));
  }
}
```

**Comment by Vampire** [comment]
Just for your information, I still don't see any timeout kick in.
I retried when I wrote my last comment, intentionally causing that deadlock, then killing the original JVM and even restarting the IDE.
The attacher process is still idling around now after 27 hours.
Where should that timeout apply?
Is it a JVM thing or a ByteBuddy thing?
Can I do something that would help improve this like providing a thread dump or similar?
